### PR TITLE
Git mapper

### LINKED ISSUE #190: Implement Git mapper (0/2)
Reference: https://oozie.apache.org/docs/5.1.0/DG_GitActionExtension.html

### PR DESCRIPTION
**Pull request by mik-laj**
Hello,

I implement a GIt Mapper. it is worth noting that the bash scripts are tested

Greets,

Closes #190

### CODE REVIEW

File: oozie-to-airflow/run-all-unit-tests [hunk 1]
```diff
@@ -16,5 +16,5 @@ set -euo pipefail
 MY_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 
 pushd "${MY_DIR}"
-pytest tests
+pytest tests scripts-tests -s
```
Comment (mik-laj) [inline oozie-to-airflow/run-all-unit-tests]: I will move this to separate PR, when I get to work.

File: oozie-to-airflow/mappers/git_mapper.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        git_branch = get_text_or_default(self.oozie_node, TAG_BRANCH, self.params)
+        destination_uri = get_text_or_default(self.oozie_node, TAG_DESTINATION_URI, self.params)
+        destination_path = urlparse(destination_uri).path
+        key_path_uri = get_text_or_default(self.oozie_node, TAG_KEY_PATH, self.params)
+        key_path = urlparse(key_path_uri).path
+        self.bash_command = prepare_git_command(
+            git_uri=git_uri, git_branch=git_branch, destination_path=destination_path, key_path=key_path
+        )
+
+    def convert_to_text(self):
```
Comment (sprzedwojski) [inline oozie-to-airflow/mappers/git_mapper.py]: It should probably be updated to match changes introduced in: https://github.com/GoogleCloudPlatform/cloud-composer/pull/167

File: oozie-to-airflow/scripts/git.sh [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+        Address to the git repository
+
+-b, --branch
+        Tree in the repository to be downloaded. Defaults to \"master\"
+
+-k, --key-path
+        An SSH key that will be used to authorize access to the repository.
+
+-d, --destination-path
+        Location on HDFS, where the repository will be upload.
```
Comment (sprzedwojski) [inline oozie-to-airflow/scripts/git.sh]:
```suggestion
        Location on HDFS, where the repository will be uploaded.
```

File: oozie-to-airflow/scripts-tests/test_git.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""""
+Tests for git action shell script
+
+Tests tests the external applications. Replaces the "gcloud" external app with a "mock" script.
```
Comment (sprzedwojski) [inline oozie-to-airflow/scripts-tests/test_git.py]:
```suggestion
Tests the external applications. Replaces the "gcloud" external app with a "mock" script.
```

File: oozie-to-airflow/scripts-tests/test_git.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        shutil.rmtree(self.app_mock_dir)
+        environ["PATH"] = self.old_path
+        return True
+
+
+class ShellScriptTestCase(unittest.TestCase):
+    """
+    Prepares the environment for script tests.
+
+    It also provides additional methods that make it easier to write tests
```
Comment (sprzedwojski) [inline oozie-to-airflow/scripts-tests/test_git.py]:
```suggestion
    It also provides additional methods that make it easier to write tests.
```

File: oozie-to-airflow/run-all-unit-tests (hunk 1 above)
Comment (mik-laj) [reply oozie-to-airflow/run-all-unit-tests]: I updated the PR.

File: oozie-to-airflow/mappers/git_mapper.py (hunk 2 above)
Comment (mik-laj) [reply oozie-to-airflow/mappers/git_mapper.py]: I rebased the PR.

File: oozie-to-airflow/scripts/git.sh (hunk 3 above)
Comment (mik-laj) [reply oozie-to-airflow/scripts/git.sh]: I updated the PR.Thanks.

File: oozie-to-airflow/scripts-tests/test_git.py (hunk 4 above)
Comment (mik-laj) [reply oozie-to-airflow/scripts-tests/test_git.py]: I updated the PR.

File: oozie-to-airflow/scripts-tests/test_git.py (hunk 5 above)
Comment (mik-laj) [reply oozie-to-airflow/scripts-tests/test_git.py]: I updated the PR.

File: oozie-to-airflow/scripts-tests/test_git.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        super().tearDown()
+        if path.isfile(self.log_file):
+            remove(self.log_file)
+        del environ["COMMAND_EXECUTION_LOG"]
+
+    def get_command_calls(self):
+        command_calls = []
+        if path.isfile(self.log_file):
+            with open(self.log_file) as file:
+                # Delete trailing the new line
```
Comment (sprzedwojski) [inline oozie-to-airflow/scripts-tests/test_git.py]:
```suggestion
                # Delete the trailing new line
```

File: oozie-to-airflow/scripts-tests/test_git.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+                file_contnet = file.read()[:-1]
```
Comment (sprzedwojski) [inline oozie-to-airflow/scripts-tests/test_git.py]:
```suggestion
                file_content = file.read()[:-1]
```

File: oozie-to-airflow/scripts-tests/test_git.py [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+                command_calls = file_contnet.split("\n")
```
Comment (sprzedwojski) [inline oozie-to-airflow/scripts-tests/test_git.py]:
```suggestion
                command_calls = file_content.split("\n")
```

File: oozie-to-airflow/utils/xml_utils.py [hunk 9]
```diff
@@ -71,3 +84,18 @@ def find_nodes_by_attribute(root, attr, val, tag=None) -> List[ET.Element]:
         if attr in node.attrib and node.attrib[attr] == val:
             matching_nodes.append(node)
     return matching_nodes
+
+
+def get_text_or_default(root: ET.Element, tag: str, params: Dict[str, str], default: str = None):
```
Comment (sprzedwojski) [inline oozie-to-airflow/utils/xml_utils.py]: I think the name of this method could be better. Currently it doesn't indicate that any EL replacing takes place inside.

Comment (mik-laj) [reply oozie-to-airflow/utils/xml_utils.py]: What do you think about `get_el_text_or_default`?

Comment (sprzedwojski) [reply oozie-to-airflow/utils/xml_utils.py]: Yeah, it's better.
I'm not sure about the `or_default` part. In general I dislike words like `and`, `or` in method names.

Maybe simply `get_tag_el_text`?

Comment (mik-laj) [reply oozie-to-airflow/utils/xml_utils.py]: I updated the PR.

File: oozie-to-airflow/scripts-tests/test_git.py (hunk 8 above)
Comment (mik-laj) [reply oozie-to-airflow/scripts-tests/test_git.py]: I updated the PR.

File: oozie-to-airflow/scripts-tests/test_git.py (hunk 7 above)
Comment (mik-laj) [reply oozie-to-airflow/scripts-tests/test_git.py]: I updated the PR. Thanks.

File: oozie-to-airflow/scripts-tests/test_git.py (hunk 6 above)
Comment (mik-laj) [reply oozie-to-airflow/scripts-tests/test_git.py]: I updated the PR. Thanks.

**Pull request merged**